ow_power_system/src/PredictionHandler.cpp

```diff
@@ -0,0 +1,137 @@
+ * a new battery EoD prediction.
+ **/
+void PredictionHandler::processMessage(const std::shared_ptr<Message>& message)
+{
+  using namespace std::chrono;
+  // The prediction handler only ever subscribes to the BatteryEoD message
+  // id, which should always be a ProgEventMessage, so this should always
+  // succeed.
+  auto prediction_msg = dynamic_cast<ProgEventMessage*>(message.get());
+  if (prediction_msg == nullptr)
+  {
+    ROS_ERROR("Failed to cast prediction message to expected type");
+    return;
+  }
+
+  // Get the event for battery EoD
+  auto eod_event = prediction_msg->getValue();
+
+  // The time of event is a `UData` structure, which represents a data
+  // point while maintaining uncertainty. For the MonteCarlo predictor
+  // used by this example, the uncertainty is captured by storing the
+  // result of each particle used in the prediction.
+  UData eod_time = eod_event.getTOE();
+  if (eod_time.uncertainty() != UType::Samples)
+  {
+    ROS_ERROR("Unexpected uncertainty type for EoD prediction");
+    return;
+  }
+
+  // valid prediction
+  // Determine the median RUL.
+  double eod_median = findMedian(eod_time.getVec());
+  auto now = MessageClock::now();
+  auto now_s = duration_cast<std::chrono::seconds>(now.time_since_epoch());
+  double rul_median = eod_median - now_s.count();
+
+  // Determine the median SOC.
+  UData currentSOC = eod_event.getState()[0];
+  double soc_median = findMedian(currentSOC.getVec());
+
+  // Determine the Battery Temperature
+  auto stateSamples = eod_event.getSystemState()[BATTERY_TEMPERATURE_INDEX];
+  std::vector<double> temperature_state;
+  for (auto sample : stateSamples)
+  {
+    temperature_state.push_back(sample[BATTERY_TEMPERATURE_INDEX]);
+  }
+
+  // HACK ALERT:
+  // The asynchronous prognoser does not have a function to get its model like
+  // the simple prognoser does. As such, one can instantiate a simple prognoser
+  // and get its model; it should be identical. However, a getter function would
+  // be much more ideal and more efficient.
+  // Construct a simple prognoser to get the battery model.
+  auto prognoser_config_path = ros::package::getPath("ow_power_system") + "/config/prognoser.cfg";
+  ConfigMap prognoser_config(prognoser_config_path);
+  std::unique_ptr<PCOE::Prognoser> temp_prog =
+    PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
+
+  auto& model = dynamic_cast<ModelBasedPrognoser*>(temp_prog.get())->getModel();
+
+  auto model_output = model.outputEqn(now_s.count(), static_cast<PrognosticsModel::state_type>(temperature_state));
+
+  // Store the newly obtained data.
+  rul_ref = rul_median;
+  soc_ref = soc_median;
+  temp_ref = model_output[MODEL_TEMPERATURE_INDEX];
+}
+
+double PredictionHandler::findMedian(std::vector<double> samples)
+{
+  std::nth_element(samples.begin(), (samples.begin() + (samples.size() / 2)),
+                    samples.end());
```

Comment:
Is there a specific guideline I should be following when it comes to indentation here? I could give each argument its own line, for example (I think I remember seeing that in the code before), but I don't recall if that's best practice or not.

I was originally just following the line character limit guidelines.

Reply:
It was a minor comment -- the second line is a space off.  (The same in my other comment).  If that's because of tab characters, those should be replaced with spaces.  Line length is the only formal guideline, though extra lines (e.g. one per argument) can sometimes make the code more readable.

Reply:
Ah, I didn't catch that. I'll go ahead and fix that, though maybe my editor is messing up the indentation because I feel I would have caught that myself while writing the code.